### ISSUE TITLE
failed: Exec failed due to IOException: Input is a directory if declare_directory passed to depset of DefaultInfo

### ISSUE THREAD

**Issue by dpoluyanov** [bazelbuild/bazel#16987]
### Description of the bug:

When directory, declared via output = ctx.actions.declare_directory passed to DefaultInfo(files=depset([output])) and used as input (srcs) of another rule, `cc_library` in my case, there is `IOException: Input is a directory` error.
It isn't happened in previous version of bazel (`6.0.0-pre.20221012.2`) and started only from 6.0.0rc1 upgrade,
all following rc candidates and `last_green` (0cc796aa2722bad60f314d1b27403dc0c30bf520) also produces such error.
It only arise when `--remote_download_minimal` is enabled, so I noted it in reproducible sample

```
ERROR: 
/private/var/tmp/_bazel_dpoluyanov/11d39f42d934765009feaa26dcc6e6cf/external/kotlin_spec_cpp/BUILD.bazel:3:17: Compiling external/kotlin_spec_cpp/kotlin-spec-cpp_grammar/UnicodeClasses.cpp failed: Exec failed due to IOException: Input is a directory: bazel-out/darwin_arm64-fastbuild/bin/external/kotlin_spec_cpp/kotlin-spec-cpp_grammar
```

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Repository with reproduce: https://github.com/dpoluyanov/bazel-gh-16987

### Which operating system are you running Bazel on?

linux, darwin

### What is the output of `bazel info release`?

release 6.0.0rc4

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
(internal repository)
```

### Have you found anything relevant by searching the web?

there is no similar issue

**Comment by fmeum** [comment]
This repros even with `--disk_cache` instead of `--remote_cache` (easier to set up for testing) and a bisect points to 32b0f5a258ee83b040d6975a6bd0795670f5cd47.

@tjgq

**Comment by tjgq** [comment]
I'm looking into it.

**Comment by meteorcloudy** [comment]
@bazel-io fork 6.0.0

**Comment by tjgq** [comment]
Adding some context here for future spelunking.

`cc_library(srcs = [tree artifact])` internally uses an action template to generate a separate action for each file inside the tree artifact. Each of these actions contains only that particular `TreeFileArtifact` in its inputs, but due to #16333, the BwoB prefetcher will attempt to fetch the entire tree artifact, which fails since the metadata for it is missing.

The temporary workaround I'm planning to submit is to fall back to including the full tree artifact in the action inputs in this situation. This might slightly degrade action performance (additional files to fetch) but unblocks the 6.0.0 release and buys us some time to fix #16333 properly and remove the workaround.

**Comment by dpoluyanov** [comment]
Thanks, @tjgq!
I've successfully rebuilt cc libraries with generated code with bazel version from 25ba76c1b8d420ebae43c1f59890ae46153c2240, so for now workaround is working as expected
